google/cloud/forseti/scanner/scanners/bigquery_scanner.py

```diff
@@ -115,32 +122,61 @@ def _retrieve(self):
 
         Returns:
             list: BigQuery ACL data
+
+        Raises:
+            ValueError: if resources have an unexpected type.
         """
         model_manager = self.service_config.model_manager
         scoped_session, data_access = model_manager.get(self.model_name)
         with scoped_session as session:
-            bq_acls = []
+            bq_acl_data = []
 
             for policy in data_access.scanner_iter(session, 'dataset_policy'):
-                dataset_id = policy.parent.name
+                # dataset_policy are always in a dataset, which is always in a
+                # project.
+                dataset = policy.parent
+                if dataset.type != 'dataset':
+                    raise ValueError(
+                        'Unexpected type of dataset_policy parent: '
+                        'got %s, want dataset' % dataset.type
+                    )
+
+                if dataset.parent.type != 'project':
+                    raise ValueError(
+                        'Unexpected type of dataset_policy grandparent: '
+                        'got %s, want project' % dataset.parent.type
+                    )
+
+                proj = project.Project(
+                    project_id=dataset.parent.name,
+                    full_name=dataset.parent.full_name,
+                    data=policy.data,
+                )
+
                 # There is no functional use for project_id in this scanner,
                 # other than to identify where the dataset comes from,
                 # which can now be done with full_name.
                 # In case you are tempted to get the project_id anyways,
                 # do not use project_id = policy.parent.parent.name
                 # which will cause db session conflict.
                 # Instead, parse the project_id from the full_name.
-                bq_acls.extend(
-                    BigqueryAccessControls.from_json(
-                        project_id=None,
-                        dataset_id=dataset_id,
-                        full_name=policy.full_name,
-                        acls=policy.data))
+                bq_acls = list(BigqueryAccessControls.from_json(
+                    project_id=None,
+                    dataset_id=dataset.name,
+                    full_name=policy.full_name,
+                    acls=policy.data))
+
+                for bq_acl in bq_acls:
+                    data = BigqueryAccessControlsData(
+                        parent_project=proj,
+                        bigquery_acl=bq_acl,
+                    )
+                    bq_acl_data.append(data)
```

Comment:
nit: You can use list comprehension here like so: (will give a little speed boost as well)

return [BigqueryAccessControlsData(parent_project=proj, bigquery_acl=acl)
            for acl in bq_acls]

Reply:
We won't be able to return here as the bq_acl_data is two levels out :( and doing 

bq_acl_data.extend([ [BigqueryAccessControlsData(parent_project=proj, bigquery_acl=acl)
for acl in bq_acls])

seems a bit harder to read, so will leave it as is for now.